### PR TITLE
Improved the parameter search space in xgboost example

### PR DESCRIPTION
**Pull request by jeromepatel**
added several new parameters in Xgboost search space which improved the accuracy of the example

## Motivation
With reference to the issue #1683, There was some need to include various parameters in <code>examples/xgboost_simple.py</code> and in<code> examples/xgboost_cv.py </code>
## Description of the changes

 I have added several parameters which include <code>eval_metric, subsample,colsample_bytree,tree_method,min_child_weight</code>
Most of the changes are in <code>examples/xgboost_simple.py</code>.
There was a significant improvement over the performance on this example, notably from 
**accuracy**:  0.9790209790209791 to
**accuracy**: 1.0 or  0.993006993006993 or 0.9860 
The inconsistency seems to appear from the random seed, but I haven't tuned the seed. Apart from that, the accuracy value would vary in range but gives higher result in just a few runs.

I have referred [xgboost docs](https://xgboost.readthedocs.io/en/latest/parameter.html) , [analytics vidhya blog](https://www.analyticsvidhya.com/blog/2016/03/complete-guide-parameter-tuning-xgboost-with-codes-python/) and pull request #1574. 
Please give me feedback if I have made any mistake or done something wrong as this is my first Pull request. 
Thank you!!

### CODE REVIEW

**Comment by jeromepatel** [comment]
Can any contributor guide me on this PR and what do I need to resolve checks?

**Comment by nzw0301** [comment]
@jeromepatel Sure. A simple way to fix the checks error is to apply `black`:

```bash
black .
```

**Comment by nzw0301** [comment]
To check them locally, [this section in `CONTRIBUTING.md`](https://github.com/optuna/optuna/blob/master/CONTRIBUTING.md#guidelines) might be helpful.

File: examples/xgboost_simple.py [hunk 1]
```diff
@@ -28,16 +30,35 @@ def objective(trial):
     param = {
         "verbosity": 0,
         "objective": "binary:logistic",
+        #use exact for small dataset
+        "tree_method": "exact",
+        #evaluation metric
+        "eval_metric": trial.suggest_categorical("eval_metric", ["rmse", "rmsle", "logloss","map"]),
```
Comment (nzw0301) [inline examples/xgboost_simple.py:36]: I'm not sure why the evaluation metric is sampled as a hyper-parameter. If the task is classification, I think that `logloss` is okay, but the others are unusual evaluation.

File: examples/xgboost_simple.py [hunk 2]
```diff
@@ -28,16 +30,35 @@ def objective(trial):
...
+        #sampling ratio for training data
+        'subsample': trial.suggest_float('subsample', 0.1,1.0),
+        #sampling according to each tree
+        'colsample_bytree': trial.suggest_float('colsample_bytree', 0.1,1.0),
     }
 
     if param["booster"] == "gbtree" or param["booster"] == "dart":
-        param["max_depth"] = trial.suggest_int("max_depth", 1, 9)
+        #maximum depth of the tree,signifies complexity of the tree
+        param["max_depth"] =  trial.suggest_int("max_depth", 3, 9)
```
Comment (nzw0301) [inline examples/xgboost_simple.py:51]: Why do you change the lower bound of `max_depth`?

Comment (nzw0301) [reply examples/xgboost_simple.py:51]: Oops, sorry it might come from [this page](https://www.analyticsvidhya.com/blog/2016/03/complete-guide-parameter-tuning-xgboost-with-codes-python/)

File: examples/xgboost_simple.py (hunk 1 above)
Comment (jeromepatel) [reply examples/xgboost_simple.py:36]: I tried directly keeping it <code> logloss </code> and it works fine, but I have also tried different <code> "eval_metric"</code> given on [this page](https://xgboost.readthedocs.io/en/latest/parameter.html), which gives about same results, so I just wanted to make sure that users are aware of available options when trying on xgboost example which can expand the search space.

**Comment by jeromepatel** [comment]
I have fixed check errors using black, and other changes were made to example, notably removed <code>"eval_metric"</code>, thanks @nzw0301 for pointing out, changed <code> "max_depth"</code> lower bound to 3 and its <code> step = 2</code>, removed <code> "min_child_weight" step</code> (both were redundant) and changed lower bound of <code> "subsample"</code> and <code>"colsample_bytree"</code>. 
The lower bound changes were made referring [here](https://www.analyticsvidhya.com/blog/2016/03/complete-guide-parameter-tuning-xgboost-with-codes-python/) and [here](https://xgboost.readthedocs.io/en/latest/parameter.html)
The performance was more consistent and didn't significantly change than previous commit.

File: examples/xgboost_simple.py [hunk 3]
```diff
@@ -28,16 +30,30 @@ def objective(trial):
...
         "alpha": trial.suggest_float("alpha", 1e-8, 1.0, log=True),
+        # sampling ratio for training data
+        "subsample": trial.suggest_float("subsample", 1e-8, 1.0),
+        # sampling according to each tree
+        "colsample_bytree": trial.suggest_float("colsample_bytree", 1e-8, 1.0),
     }
 
     if param["booster"] == "gbtree" or param["booster"] == "dart":
-        param["max_depth"] = trial.suggest_int("max_depth", 1, 9)
+        # maximum depth of the tree,signifies complexity of the tree
```
Comment (nzw0301) [inline examples/xgboost_simple.py:48]: it needs one space.

```suggestion
        # maximum depth of the tree, signifies complexity of the tree
```

File: examples/xgboost_simple.py [hunk 4]
```diff
@@ -19,6 +19,8 @@
 
 # FYI: Objective functions can take additional arguments
 # (https://optuna.readthedocs.io/en/stable/faq.html#objective-func-additional-args).
+
+
```
Comment (nzw0301) [inline examples/xgboost_simple.py:23]: Could you remove these blank lines because the comments above are related to the function?

File: examples/xgboost_simple.py (hunk 3 above)
Comment (jeromepatel) [reply examples/xgboost_simple.py:48]: okay, I'll add it!!

File: examples/xgboost_simple.py (hunk 4 above)
Comment (jeromepatel) [reply examples/xgboost_simple.py:23]: Okay, I'll remove that!! Thanks for your review!!

**Comment by jeromepatel** [comment]
I have removed space and extra lines in <code>examples/xgboost_simple.py</code> as told @nzw0301, Thank you for your valuable time!!

**Comment by jeromepatel** [comment]
I only have changed two things as mention above, but I have no idea why this particular test failed, it was working on the previous commit. If you find anything helpful please guide me on this issue!

**Comment by toshihikoyanase** [comment]
@c-bata Would you review this PR since you create #1683? If you don't have time enough, please feel free to remove the assignment.

CC @g-votte @smly If you have any comments about the search space and training settings, please let us know.

**Comment by c-bata** [comment]
Could you assign another committers for this PR because I'm not familiar with XGBoost very much?

**Comment by toshihikoyanase** [comment]
@c-bata Sure! Thank you for letting me know. I'll review it with the help of data scientists.

**Comment by jeromepatel** [comment]
@toshihikoyanase Can you check my latest commit, I have made cosmetic changes to pass the test, and apart from that most of things are described above.

**Comment by jeromepatel** [comment]
@toshihikoyanase I have some doubt regarding this PR, I hope some guidance will clear my doubts. Actually, this is my first PR and it is still open, I wanted to work on a different issue(sampler update) as I guess this PR might not need further major modification. So, can I work on different issues at the same time? and also if I create another branch to work on other issues, would it affect anything on this branch(with this PR)? I am a bit new to all of this, so sorry if my question was silly. Btw, thanks for helping me out!!

**Comment by toshihikoyanase** [comment]
@jeromepatel First of all, sorry for my delayed response. I'm taking days off this week, and I'll review this PR next week.

> can I work on different issues at the same time?

Thank you for your hard work. Your further contribution is more than welcome. I saw your comment in #2233, and I guess you'll update one of the samplers. I think it is a good choice because you can refer to the fix for `PartialFixedSampler` (#2209).

> if I create another branch to work on other issues, would it affect anything on this branch(with this PR)? 

If you create another branch from the `master` branch, the two branches do not affect each other. So, please make sure to check out `master` before creating your new feature branch.

**Comment by jeromepatel** [comment]
@toshihikoyanase  Thank you so much for your quick reply and help. It cleared all my doubts. And indeed I am working on issue #2233, again thank you for your contributions. I am grateful to work in this awesome community!!

**Review by toshihikoyanase — COMMENTED** [review]
Thank you for your PR. I added some early comments.
Next, I'd like to investigate the behavior of the search algorithm using Optuna's [visualization features](https://optuna.readthedocs.io/en/stable/reference/visualization/index.html). At least, I'd like to measure the [importance](https://optuna.readthedocs.io/en/stable/reference/visualization/generated/optuna.visualization.plot_param_importances.html#optuna.visualization.plot_param_importances) of the hyperparameters so that we can focus on the important area of the search space.

BTW, I noticed the dataset of this example has only 569 samples, and I guess it may be a bit too small. Do you have any idea of alternative datasets?

**Comment by toshihikoyanase** [comment]
Oops, I overlooked you used your master branch to create this PR. Then, the procedure that I suggested is not enough to keep two branches clean.
In this case, I guess you need to create another local branch that follows the commits in `optuna/optuna`'s `master` branch.

I assume that you set the remote url `git@github.com:optuna/optuna.git` as `upstream` (i.e., you executed `git remote add  git@github.com:optuna/optuna.git`).
Firstly, please create the `upstream-master` branch as follows:
```
$ git fetch upstream
$ git checkout -b upstream-master upstream/master
```
Secondly, please check out your feature branch from `upstream-master`:
```
$ git checkout -b your-feature-branch
```
Finally, please push the `your-feature-branch` to `origin` to create a new PR.
```
$ git push origin your-feature-branch
```

If you have any questions, please let me know.

(I usually use my local `master` branch to just follow the change in `optuna/optuna`'s master, so I can simply check out the feature branch from my `master` branch. Sorry for bothering you.)

File: examples/xgboost_cv.py [hunk 5]
```diff
@@ -35,10 +35,16 @@ def objective(trial):
         "booster": trial.suggest_categorical("booster", ["gbtree", "gblinear", "dart"]),
         "lambda": trial.suggest_loguniform("lambda", 1e-8, 1.0),
         "alpha": trial.suggest_loguniform("alpha", 1e-8, 1.0),
+        # sampling ratio for training data
+        "subsample": trial.suggest_float("subsample", 1e-8, 1.0),
```
[the remark this answers is not in the document]
Comment (jeromepatel) [reply examples/xgboost_cv.py:39]: Okay, I'll modify that, using <code>low=0.2</code> would be fine I guess ( I also have looked into <code>best_trial subsample</code> value which generally ranges above 0.5). Previously, I thought that lowering <code>subsample</code> might increase the search space( regardless of dataset size, as some user might use this example ranges directly)

File: examples/xgboost_cv.py [hunk 6]
```diff
@@ -35,10 +35,16 @@ def objective(trial):
...
+        # sampling according to each tree
+        "colsample_bytree": trial.suggest_float("colsample_bytree", 1e-8, 1.0),
```
[the remark this answers is not in the document]
Comment (jeromepatel) [reply examples/xgboost_cv.py:41]: Okay, I'll change that too.

File: examples/xgboost_cv.py [hunk 7]
```diff
@@ -35,10 +35,16 @@ def objective(trial):
...
     }
 
     if param["booster"] == "gbtree" or param["booster"] == "dart":
         param["max_depth"] = trial.suggest_int("max_depth", 1, 9)
+        # minimum child weight, larger the term more conservative the tree
+        param["min_child_weight"] = trial.suggest_int("min_child_weight", 3, 10)
```
[the remark this answers is not in the document]
Comment (jeromepatel) [reply examples/xgboost_cv.py:47]: As the dataset is small with only 569 samples, there are higher chances to overfit for the model with a default value of <code>min_child_weight=1</code>, in my experimentation with parameter search/tuning, I have found that with increasing <code>min_child_weight</code>, it regularizes the algorithm. So, I have used value of 3, you can find further explanation [here](https://stats.stackexchange.com/questions/317073/explanation-of-min-child-weight-in-xgboost-algorithm) that why <code>min_child_weight</code> helps in regularization.

**Comment by jeromepatel** [comment]
Thank you for guiding me through this branch resolution, I will follow the above steps, make suggested changes to the example and get back to you. 

> BTW, I noticed the dataset of this example has only 569 samples, and I guess it may be a bit too small. Do you have any idea of alternative datasets?
[Here](https://machinelearningmastery.com/standard-machine-learning-datasets/) are some standard datasets, I suggest using Wine Quality Dataset (2nd dataset in the link) which contains <code> 4,898 samples with 11 features and 1 target variable</code>.

**Comment by jeromepatel** [comment]
First of all, sorry for any inconvenience that occurred due to me, I have successfully migrated changes to another branch from master, the corresponding PR Pull #2346. It contains changes in two files <code>xgboost_example.</code>
Thank you for your patience.  I am closing this pull request now.
